File: src/internalDebug/java/cash/andrew/mntrailconditions/DebugMnTrailConditionsModule.java
package cash.andrew.mntrailconditions;

import cash.andrew.mntrailconditions.data.DebugDataModule;
import cash.andrew.mntrailconditions.ui.DebugUiModule;
import com.facebook.stetho.Stetho;
import com.facebook.stetho.timber.StethoTree;
import dagger.Module;
import dagger.Provides;
import javax.inject.Singleton;
import timber.log.Timber;

@Module(
    addsTo = MnTrailConditionsModule.class,
    includes = {
        DebugUiModule.class,
        DebugDataModule.class
    },
    overrides = true
)
public final class DebugMnTrailConditionsModule {
  // Low-tech flag to force certain debug build behaviors when running in an instrumentation test.
  // This value is used in the creation of singletons so it must be set before the graph is created.
  static boolean instrumentationTest = false;

  @Provides @Singleton @IsInstrumentationTest boolean provideIsInstrumentationTest() {
    return instrumentationTest;
  }

  @Provides @Singleton MnTrailConditionsInitializer provideMnTrailConditionsInitializer() {
    return context -> {
      Stetho.initializeWithDefaults(context);

      Timber.plant(new Timber.DebugTree());
      Timber.plant(new StethoTree());
    };
  }
}

File: src/internalDebug/kotlin/cash/andrew/mntrailconditions/DebugMnTrailConditionsModule.kt
package cash.andrew.mntrailconditions

import android.app.Application
import cash.andrew.mntrailconditions.data.DebugDataModule
import com.crashlytics.android.Crashlytics
import com.crashlytics.android.core.CrashlyticsCore
import com.facebook.stetho.Stetho
import com.facebook.stetho.timber.StethoTree
import com.google.firebase.messaging.FirebaseMessaging
import dagger.Module
import dagger.Provides
import io.fabric.sdk.android.Fabric
import timber.log.Timber
import android.os.StrictMode



@Module(includes = [DebugDataModule::class])
object DebugMnTrailConditionsModule {
    @JvmStatic
    @Provides
    fun provideMnTrailConditionsInitializer(
            app: Application,
            firebaseMessaging: FirebaseMessaging
    ): MnTrailConditionsInitializer = { context ->
        Stetho.initializeWithDefaults(context)

        Timber.plant(Timber.DebugTree())
        Timber.plant(StethoTree())

        // Set up Crashlytics, disabled for debug builds
        val crashlyticsKit = Crashlytics.Builder()
                .core(CrashlyticsCore.Builder().disabled(true).build())
                .build()

        // Initialize Fabric with the debug-disabled crashlytics.
        Fabric.with(app, crashlyticsKit)

        StrictMode.setThreadPolicy(StrictMode.ThreadPolicy.Builder()
                .detectAll()
                .penaltyLog()
                .penaltyDeath()
                .build())

        StrictMode.setVmPolicy(StrictMode.VmPolicy.Builder()
                .detectAll()
                .penaltyLog()
                .penaltyDeath()
                .build())

        // allows updating test in firestore to get push notifications
        // and ensure everything is working smoothly.
        firebaseMessaging.subscribeToTopic("test")
    }
}
